docs/source/notes/gradcheck.rst

```diff
@@ -0,0 +1,218 @@
+.. _gradcheck-mechanics:
+
+Gradcheck mechanics
+===================
+
+This note presents an overview of how the :meth:`~torch.autograd.gradcheck` and :meth:`~torch.autograd.gradgradcheck` functions work.
+
+It will cover both forward and backward mode AD as well as complex valued functions and higher order derivatives.
+
+.. contents:: :local:
+    :depth: 2
+
+Notations and background informations
+-------------------------------------
+
+Throughout this note, we will use :math:`x`, :math:`y`, :math:`a`, :math:`b`, :math:`v`, :math:`u`, :math:`ur` and :math:`ui` as real valued vectors and :math:`z` is a complex valued vector.
+:math:`N` and :math:`M` are two integers that we will use for the dimension of the input and output space respectively.
+
+We will use :math:`f: \mathcal{R}^N \to \mathcal{R}^M` as our basic real to real function such that :math:`y = f(x)`.
+We will use :math:`g: \mathcal{C}^N \to \mathcal{R}^M` as our basic complex to real function such that :math:`y = g(z)`.
+We will also use :math:`z = a + i b`.
+
+
+For the simple real to real case, we write as :math:`J_f` the jacobian matrix associated with :math:`f` of size :math:`M \times N`.
+This matrix contains all the partial derivatives such that the entry at position :math:`(i, j)` contains :math:`\frac{\partial y_i}{\partial x_j}`.
+Backward mode AD is then computing, for a given vector :math:`v` of size :math:`M`, the quantity :math:`v^T J_f`.
+Forward mode AD on the other hand is computing, for a given vector :math:`u` of size :math:`N`, the quantity :math:`J_f u`.
+
+For functions that contain complex values, the story is a lot more complex. We only provide the gist here and the full description can be found at :ref:`complex_autograd-doc`.
+
+Proper complex derivatives are too restrictive for most functions we care about. So we turn to Wirtinger calculus.
+In a basic setting of Wirtinger calculus, the chain rule required access to both the Wirtinger derivative (called :math:`W` below) and the Conjugate Wirtinger derivative (called :math:`CW` below). This means that in the general case of Wirtinger calculus, the values that we would need to "backward through the graph" in backward mode AD would need to be twice the size of the "forward value".
+
+To avoid this problem, for backward mode AD, we always work under the assumption that the current function is part of a bigger function whose output is in :math:`\mathcal{R}`. Note that this assumption means that all the intermediary gradients we compute during the backward pass are also corresponding to functions whose output is in :math:`\mathcal{R}`.
```

Comment:
Maybe also add: "we can assume the overall function is R-valued because thats all we need for the purposes of optimization."